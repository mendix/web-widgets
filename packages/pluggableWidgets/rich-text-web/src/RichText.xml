<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.custom.richtext.RichText" needsEntityContext="true" pluginWidget="true" offlineCapable="true" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../node_modules/mendix/custom_widget.xsd">
    <name>Rich Text</name>
    <description>Rich inline or toolbar text editing</description>
    <studioProCategory>Input elements</studioProCategory>
    <studioCategory>Input Elements</studioCategory>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="Data source">
                <property key="stringAttribute" type="attribute">
                    <caption>Value attribute</caption>
                    <description>The attribute used for the content of the text editor, recommendation is to use an unlimited string data type.</description>
                    <attributeTypes>
                        <attributeType name="String" />
                    </attributeTypes>
                </property>
            </propertyGroup>
            <propertyGroup caption="General">
                <systemProperty key="Label" />
                <property key="menubarMode" type="enumeration" defaultValue="hide">
                    <caption>Menu bar</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="hide">Hidden</enumerationValue>
                        <enumerationValue key="basic">Basic</enumerationValue>
                        <enumerationValue key="full">Full</enumerationValue>
                        <enumerationValue key="custom">Custom</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="enableStatusBar" type="boolean" defaultValue="true">
                    <caption>Enable status bar</caption>
                    <description />
                </property>
            </propertyGroup>
            <propertyGroup caption="Toolbar">
                <property key="preset" type="enumeration" defaultValue="basic">
                    <caption>Toolbar</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="basic">Basic</enumerationValue>
                        <enumerationValue key="standard">Standard</enumerationValue>
                        <enumerationValue key="full">Full</enumerationValue>
                        <enumerationValue key="custom">Custom</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="toolbarMode" type="enumeration" defaultValue="sliding">
                    <caption>Mode</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="sliding">Sliding</enumerationValue>
                        <enumerationValue key="floating">Floating</enumerationValue>
                        <enumerationValue key="scrolling">Scrolling</enumerationValue>
                        <enumerationValue key="wrap">Wrap</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="toolbarLocation" type="enumeration" defaultValue="auto">
                    <caption>Location</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="auto">Auto</enumerationValue>
                        <enumerationValue key="top">Top</enumerationValue>
                        <enumerationValue key="bottom">Bottom</enumerationValue>
                        <enumerationValue key="inline">Inline</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="quickbars" type="boolean" defaultValue="false">
                    <caption>Enable quick toolbar</caption>
                    <description />
                </property>
                <property key="contextmenutype" type="enumeration" defaultValue="native">
                    <caption>Context menu</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="native">Native</enumerationValue>
                        <enumerationValue key="richtext">Rich Text</enumerationValue>
                    </enumerationValues>
                </property>
            </propertyGroup>
            <propertyGroup caption="Editability">
                <systemProperty key="Editability" />
                <property key="readOnlyStyle" type="enumeration" defaultValue="text">
                    <caption>Read-only style</caption>
                    <description>How the rich text editor will appear in read-only mode.</description>
                    <enumerationValues>
                        <enumerationValue key="text">Rich text</enumerationValue>
                        <enumerationValue key="bordered">Bordered</enumerationValue>
                        <enumerationValue key="readPanel">Read panel</enumerationValue>
                    </enumerationValues>
                </property>
            </propertyGroup>
            <propertyGroup caption="Visibility">
                <systemProperty key="Visibility" />
            </propertyGroup>
        </propertyGroup>
        <propertyGroup caption="Dimensions">
            <propertyGroup caption="Dimensions">
                <property key="widthUnit" type="enumeration" defaultValue="percentage">
                    <caption>Width unit</caption>
                    <description>Percentage: portion of parent size. Pixels: absolute amount of pixels.</description>
                    <enumerationValues>
                        <enumerationValue key="percentage">Percentage</enumerationValue>
                        <enumerationValue key="pixels">Pixels</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="width" type="integer" defaultValue="100">
                    <caption>Width</caption>
                    <description />
                </property>
                <property key="heightUnit" type="enumeration" defaultValue="pixels">
                    <caption>Height unit</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="percentageOfWidth">Percentage of width</enumerationValue>
                        <enumerationValue key="pixels">Pixels</enumerationValue>
                        <enumerationValue key="percentageOfParent">Percentage of parent</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="height" type="integer" defaultValue="250">
                    <caption>Height</caption>
                    <description />
                </property>
                <property key="minHeight" type="integer" defaultValue="250">
                    <caption>Minimum height</caption>
                    <description>Editor's minimum height. The number value is in pixels.</description>
                </property>
                <property key="resize" type="enumeration" defaultValue="false">
                    <caption>Editor resize</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="false">No</enumerationValue>
                        <enumerationValue key="true">Vertical</enumerationValue>
                        <enumerationValue key="both">Both</enumerationValue>
                    </enumerationValues>
                </property>
            </propertyGroup>
        </propertyGroup>
        <propertyGroup caption="Events">
            <propertyGroup caption="Events">
                <property key="onChange" type="action">
                    <caption>On change</caption>
                    <description />
                </property>
                <property key="onFocus" type="action">
                    <caption>On enter</caption>
                    <description />
                </property>
                <property key="onBlur" type="action">
                    <caption>On leave</caption>
                    <description />
                </property>
                <property key="onKeyPress" type="action">
                    <caption>On key press</caption>
                    <description />
                </property>
            </propertyGroup>
            <propertyGroup caption="On change behavior">
                <property key="onChangeType" type="enumeration" defaultValue="onLeave">
                    <caption>Editor resize</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="onLeave">When user leaves input field</enumerationValue>
                        <enumerationValue key="onDataChange">While user is entering data</enumerationValue>
                    </enumerationValues>
                </property>
            </propertyGroup>
        </propertyGroup>
        <propertyGroup caption="Advanced">
            <propertyGroup caption="Advanced">
                <property key="extended_valid_elements" type="textTemplate" required="false">
                    <caption>Additional valid elements</caption>
                    <description>Additional sanitization rules to allow certain tag, sample: script[src|async]</description>
                </property>
                <property key="spellCheck" type="boolean" defaultValue="false">
                    <caption>Enable spell checking</caption>
                    <description />
                </property>
                <property key="highlight_on_focus" type="boolean" defaultValue="false">
                    <caption>Highlight on focus</caption>
                    <description />
                </property>
                <property key="content_css" type="textTemplate" required="false">
                    <caption>Additional stylesheet</caption>
                    <description>Additional stylesheet to be included in the rich text content</description>
                </property>
            </propertyGroup>
        </propertyGroup>

        <propertyGroup caption="Custom toolbar">
            <propertyGroup caption="Custom toolbar">
                <property key="toolbarConfig" type="enumeration" defaultValue="basic">
                    <caption>Toolbar group</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="basic">Basic</enumerationValue>
                        <enumerationValue key="advanced">Advanced</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="basicstyle" type="boolean" defaultValue="true">
                    <caption>Basic style</caption>
                    <description />
                </property>
                <property key="extendedstyle" type="boolean" defaultValue="true">
                    <caption>Extended style</caption>
                    <description />
                </property>
                <property key="textalign" type="boolean" defaultValue="true">
                    <caption>Text alignment</caption>
                    <description />
                </property>
                <property key="clipboard" type="boolean" defaultValue="true">
                    <caption>Clipboard</caption>
                    <description />
                </property>
                <property key="fontstyle" type="boolean" defaultValue="true">
                    <caption>Font style</caption>
                    <description />
                </property>
                <property key="paragraph" type="boolean" defaultValue="true">
                    <caption>Paragraph</caption>
                    <description />
                </property>
                <property key="document" type="boolean" defaultValue="true">
                    <caption>Document</caption>
                    <description />
                </property>
                <property key="history" type="boolean" defaultValue="true">
                    <caption>History</caption>
                    <description />
                </property>
                <property key="accordion" type="boolean" defaultValue="true">
                    <caption>Accordion</caption>
                    <description />
                </property>
                <property key="code" type="boolean" defaultValue="true">
                    <caption>Code</caption>
                    <description />
                </property>
                <property key="anchor" type="boolean" defaultValue="true">
                    <caption>Anchor</caption>
                    <description />
                </property>
                <property key="direction" type="boolean" defaultValue="true">
                    <caption>Text direction</caption>
                    <description />
                </property>
                <property key="link" type="boolean" defaultValue="true">
                    <caption>Link</caption>
                    <description />
                </property>
                <property key="list" type="boolean" defaultValue="true">
                    <caption>List</caption>
                    <description />
                </property>
                <property key="preview" type="boolean" defaultValue="true">
                    <caption>Preview</caption>
                    <description />
                </property>
                <property key="table" type="boolean" defaultValue="true">
                    <caption>Table</caption>
                    <description />
                </property>
                <property key="visualaid" type="boolean" defaultValue="true">
                    <caption>Visual aid</caption>
                    <description />
                </property>
                <property key="media" type="boolean" defaultValue="true">
                    <caption>Media</caption>
                    <description />
                </property>
                <property key="util" type="boolean" defaultValue="true">
                    <caption>Utility</caption>
                    <description />
                </property>
                <property key="emoticon" type="boolean" defaultValue="true">
                    <caption>Emoticon</caption>
                    <description />
                </property>
                <property key="remove" type="boolean" defaultValue="true">
                    <caption>Removal</caption>
                    <description />
                </property>
                <property key="advancedConfig" type="object" isList="true" required="false">
                    <caption>Advanced groups</caption>
                    <description />
                    <properties>
                        <property key="ctItemType" type="enumeration" required="true" defaultValue="separator">
                            <caption>Button</caption>
                            <category>Item</category>
                            <description>Button Type</description>
                            <enumerationValues>
                                <enumerationValue key="separator">separator</enumerationValue>
                                <enumerationValue key="aligncenter">aligncenter</enumerationValue>
                                <enumerationValue key="alignjustify">alignjustify</enumerationValue>
                                <enumerationValue key="alignleft">alignleft</enumerationValue>
                                <enumerationValue key="alignnone">alignnone</enumerationValue>
                                <enumerationValue key="alignright">alignright</enumerationValue>
                                <enumerationValue key="blockquote">blockquote</enumerationValue>
                                <enumerationValue key="backcolor">backcolor</enumerationValue>
                                <enumerationValue key="blocks">blocks</enumerationValue>
                                <enumerationValue key="bold">bold</enumerationValue>
                                <enumerationValue key="copy">copy</enumerationValue>
                                <enumerationValue key="cut">cut</enumerationValue>
                                <enumerationValue key="fontfamily">fontfamily</enumerationValue>
                                <enumerationValue key="fontsize">fontsize</enumerationValue>
                                <enumerationValue key="forecolor">forecolor</enumerationValue>
                                <enumerationValue key="hr">hr</enumerationValue>
                                <enumerationValue key="indent">indent</enumerationValue>
                                <enumerationValue key="italic">italic</enumerationValue>
                                <enumerationValue key="lineheight">lineheight</enumerationValue>
                                <enumerationValue key="newdocument">newdocument</enumerationValue>
                                <enumerationValue key="outdent">outdent</enumerationValue>
                                <enumerationValue key="paste">paste</enumerationValue>
                                <enumerationValue key="pastetext">pastetext</enumerationValue>
                                <enumerationValue key="print">print</enumerationValue>
                                <enumerationValue key="redo">redo</enumerationValue>
                                <enumerationValue key="remove">remove</enumerationValue>
                                <enumerationValue key="removeformat">removeformat</enumerationValue>
                                <enumerationValue key="selectall">selectall</enumerationValue>
                                <enumerationValue key="strikethrough">strikethrough</enumerationValue>
                                <enumerationValue key="subscript">subscript</enumerationValue>
                                <enumerationValue key="superscript">superscript</enumerationValue>
                                <enumerationValue key="underline">underline</enumerationValue>
                                <enumerationValue key="undo">undo</enumerationValue>
                                <enumerationValue key="visualaid">visualaid</enumerationValue>
                                <enumerationValue key="accordion">accordion</enumerationValue>
                                <enumerationValue key="code">code</enumerationValue>
                                <enumerationValue key="anchor">anchor</enumerationValue>
                                <enumerationValue key="charmap">charmap</enumerationValue>
                                <enumerationValue key="codesample">codesample</enumerationValue>
                                <enumerationValue key="ltr">ltr</enumerationValue>
                                <enumerationValue key="rtl">rtl</enumerationValue>
                                <enumerationValue key="emoticons">emoticons</enumerationValue>
                                <enumerationValue key="fullscreen">fullscreen</enumerationValue>
                                <enumerationValue key="help">help</enumerationValue>
                                <enumerationValue key="image">image</enumerationValue>
                                <enumerationValue key="insertdatetime">insertdatetime</enumerationValue>
                                <enumerationValue key="link">link</enumerationValue>
                                <enumerationValue key="openlink">openlink</enumerationValue>
                                <enumerationValue key="unlink">unlink</enumerationValue>
                                <enumerationValue key="bullist">bullist</enumerationValue>
                                <enumerationValue key="numlist">numlist</enumerationValue>
                                <enumerationValue key="media">media</enumerationValue>
                                <enumerationValue key="pagebreak">pagebreak</enumerationValue>
                                <enumerationValue key="preview">preview</enumerationValue>
                                <enumerationValue key="searchreplace">searchreplace</enumerationValue>
                                <enumerationValue key="table">table</enumerationValue>
                                <enumerationValue key="tabledelete">tabledelete</enumerationValue>
                                <enumerationValue key="tableinsertdialog">tableinsertdialog</enumerationValue>
                                <enumerationValue key="visualblocks">visualblocks</enumerationValue>
                                <enumerationValue key="visualchars">visualchars</enumerationValue>
                                <enumerationValue key="wordcount">wordcount</enumerationValue>
                            </enumerationValues>
                        </property>
                    </properties>
                </property>
            </propertyGroup>
        </propertyGroup>
        <propertyGroup caption="Custom menubar">
            <propertyGroup caption="Custom menubar">
                <property key="menubarConfig" type="enumeration" defaultValue="basic">
                    <caption>Toolbar group</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="basic">Basic</enumerationValue>
                        <enumerationValue key="advanced">Advanced</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="fileMenubar" type="boolean" defaultValue="true">
                    <caption>File</caption>
                    <description />
                </property>
                <property key="editMenubar" type="boolean" defaultValue="true">
                    <caption>Edit</caption>
                    <description />
                </property>
                <property key="insertMenubar" type="boolean" defaultValue="true">
                    <caption>Insert</caption>
                    <description />
                </property>
                <property key="viewMenubar" type="boolean" defaultValue="true">
                    <caption>View</caption>
                    <description />
                </property>
                <property key="formatMenubar" type="boolean" defaultValue="true">
                    <caption>Format</caption>
                    <description />
                </property>
                <property key="tableMenubar" type="boolean" defaultValue="true">
                    <caption>Table</caption>
                    <description />
                </property>
                <property key="toolsMenubar" type="boolean" defaultValue="true">
                    <caption>Tools</caption>
                    <description />
                </property>
                <property key="helpMenubar" type="boolean" defaultValue="true">
                    <caption>Help</caption>
                    <description />
                </property>
                <property key="advancedMenubarConfig" type="object" isList="true" required="false">
                    <caption>Advanced groups</caption>
                    <description />
                    <properties>
                        <property key="menubarItemType" type="enumeration" required="true" defaultValue="file">
                            <caption>Button</caption>
                            <category>Item</category>
                            <description>Button Type</description>
                            <enumerationValues>
                                <enumerationValue key="file">File</enumerationValue>
                                <enumerationValue key="edit">Edit</enumerationValue>
                                <enumerationValue key="insert">Insert</enumerationValue>
                                <enumerationValue key="view">View</enumerationValue>
                                <enumerationValue key="format">Format</enumerationValue>
                                <enumerationValue key="table">Table</enumerationValue>
                                <enumerationValue key="tools">Tools</enumerationValue>
                                <enumerationValue key="help">Help</enumerationValue>
                            </enumerationValues>
                        </property>
                    </properties>
                </property>
            </propertyGroup>
        </propertyGroup>
    </properties>
</widget>
