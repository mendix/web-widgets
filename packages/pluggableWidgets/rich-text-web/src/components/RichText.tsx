import { createElement, useState } from "react";
import { getDimensions } from "@mendix/pluggable-widgets-commons";
import { RichTextContainerProps } from "../../typings/RichTextProps";
import loadingCircleSvg from "../ui/loading-circle.svg";
import classNames from "classnames";
import { Editor } from "./Editor";

export function RichText(props: RichTextContainerProps): JSX.Element | null {
    const { width: w, height: h, widthUnit, heightUnit, readOnlyStyle, name, id } = props;
    const [element, setElement] = useState<HTMLElement | null>(null);

    if (props.stringAttribute.status !== "available") {
        return <img src={loadingCircleSvg} className="widget-rich-text-loading-spinner" alt="" aria-hidden />;
    }

    const { width, height } = getDimensions({
        width: w,
        widthUnit,
        height: h,
        heightUnit
    });

    const readOnly = props.stringAttribute.readOnly;

    return (
        <div
            id={id}
            className={classNames("widget-rich-text", `${readOnly ? `editor-${readOnlyStyle}` : ""}`)}
            style={{ width, height }}
        >
            <div ref={setElement} id={name} />
            {element ? <Editor element={element} widgetProps={props} /> : null}
        </div>
    );
}
