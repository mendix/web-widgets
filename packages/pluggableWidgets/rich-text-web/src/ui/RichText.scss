$rte-brand-primary: #264ae5;
.widget-rich-text {
    overflow-y: auto;
    display: flex;
    flex-direction: column;
    .tox-tinymce {
        border-radius: 4px;
        border-width: 1px;
        flex-grow: 1;
    }

    .tox .tox-edit-area::before {
        border-color: $rte-brand-primary;
        border-radius: 1px;
        border-width: 1px;
    }

    &.editor-bordered,
    &.editor-text {
        .tox-edit-area__iframe {
            background-color: transparent;
        }
    }

    &.editor-text {
        border: unset;

        .tox-tinymce {
            border: unset;
        }
    }

    .mce-content-body {
        width: 100%;
        outline: 0;
    }
}

.modal-dialog {
    & ~ .tox-tinymce-inline {
        &.tox-tinymce--toolbar-sticky-off {
            z-index: 102;
        }
    }
}
