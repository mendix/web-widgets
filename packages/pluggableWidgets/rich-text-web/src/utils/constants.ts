function getBaseUrl(): string {
    const base_url = document.getElementsByTagName("base");
    return base_url && base_url.length > 0 ? base_url[0]?.href : "";
}
const PLUGINS =
    "accordion advlist anchor autolink charmap code codesample directionality emoticons fullscreen image help importcss insertdatetime link lists media nonbreaking pagebreak preview quickbars searchreplace table visualblocks visualchars wordcount";

export const DEFAULT_CONFIG = {
    width: "100%",
    promotion: false,
    branding: false,
    skin: "oxide",
    content_css: "default",
    base_url: `${getBaseUrl()}widgets/com/mendix/widget/custom/richtext`,
    plugins: PLUGINS,
    font_size_formats: "8px 10px 12px 14px 18px 24px 36px",
    menu: {
        file: { title: "File", items: "newdocument restoredraft | preview | export print | deleteallconversations" },
        edit: { title: "Edit", items: "undo redo | cut copy paste pastetext | selectall | searchreplace" },
        view: {
            title: "View",
            items: "code | visualaid visualchars visualblocks | spellchecker | preview fullscreen | showcomments"
        },
        insert: {
            title: "Insert",
            items: "image link media addcomment pageembed codesample inserttable | charmap emoticons hr | pagebreak nonbreaking anchor tableofcontents | insertdatetime"
        },
        format: {
            title: "Format",
            items: "bold italic underline strikethrough superscript subscript codeformat | blocks fontfamily fontsize align lineheight | forecolor backcolor | language | removeformat"
        },
        tools: { title: "Tools", items: "spellchecker spellcheckerlanguage | a11ycheck code wordcount" },
        table: { title: "Table", items: "inserttable | cell row column | advtablesort | tableprops deletetable" }
    }
};
