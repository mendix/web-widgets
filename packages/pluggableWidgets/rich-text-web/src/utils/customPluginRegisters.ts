import Quill from "quill";
import MendixTheme from "./themes/mxTheme";
import "./formats/fonts";
import "./formats/fontsize";
import CustomListItem from "./formats/customList";
import CustomLink from "./formats/link";
import CustomVideo from "./formats/video";
import { Attributor } from "parchment";
const direction = Quill.import("attributors/style/direction") as Attributor;
const alignment = Quill.import("attributors/style/align") as Attributor;
import { IndentLeftStyle, IndentRightStyle } from "./formats/indent";
import Formula from "./formats/formula";
import CustomHeader from "./formats/header";
class Empty {
    doSomething(): string {
        return "";
    }
}

/**
 * Custom format registration for quill.
 */
Quill.debug("error");
Quill.register({ "themes/snow": MendixTheme }, true);
Quill.register(CustomListItem, true);
Quill.register(CustomLink, true);
Quill.register(CustomVideo, true);
Quill.register(direction, true);
Quill.register(alignment, true);
Quill.register(IndentLeftStyle, true);
Quill.register(IndentRightStyle, true);
Quill.register(Formula, true);
Quill.register(CustomHeader, true);
// add empty handler for view code, this format is handled by toolbar's custom config via ViewCodeDialog
Quill.register({ "ui/view-code": Empty });
