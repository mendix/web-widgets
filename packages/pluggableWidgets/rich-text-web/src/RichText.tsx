import { ValidationAlert } from "@mendix/widget-plugin-component-kit/Alert";
import { getDimensions } from "@mendix/widget-plugin-platform/utils/get-dimensions";
import classNames from "classnames";
import { createElement, Fragment } from "react";
import { RichTextContainerProps } from "../typings/RichTextProps";
import BundledEditor from "./components/Editor";
import "./ui/RichText.scss";
import { constructWrapperStyle } from "./utils/helpers";
import { createMenubar } from "./utils/menubar";
import { createPreset } from "./utils/presets";

export default function RichText(props: RichTextContainerProps): JSX.Element {
    const {
        stringAttribute,
        id,
        width: w,
        height: h,
        widthUnit,
        heightUnit,
        preset,
        menubarMode,
        readOnlyStyle,
        enableStatusBar,
        resize
    } = props;

    const { width, height } = getDimensions({
        width: w,
        widthUnit,
        height: h,
        heightUnit
    });
    const wrapperAttributes = stringAttribute?.readOnly && readOnlyStyle !== "readPanel" ? { readOnly: true } : {};

    const presets = createPreset(preset, props);
    const menubar = createMenubar(menubarMode, props);
    const wrapperStyle = constructWrapperStyle(props, { width, height });
    return (
        <Fragment>
            <div
                id={id}
                className={classNames(
                    "widget-rich-text",
                    `${stringAttribute?.readOnly ? `editor-${readOnlyStyle}` : ""}`,
                    {
                        "form-control": props.toolbarLocation === "inline",
                        "widget-rich-text-min-height": heightUnit !== "pixels" && !stringAttribute?.readOnly,
                        "widget-rich-text-min-height-readonly": heightUnit !== "pixels" && stringAttribute?.readOnly
                    }
                )}
                style={wrapperStyle}
                {...wrapperAttributes}
            >
                {stringAttribute.status === "loading" || stringAttribute.status !== "available" ? (
                    <div className="mx-progress"></div>
                ) : (
                    <BundledEditor
                        {...props}
                        menubar={menubar}
                        toolbar={presets.toolbar}
                        editorHeight={height}
                        editorWidth={width}
                        key={`${String(stringAttribute.readOnly)}_${id}_${props.content_css?.value}`}
                        resize={enableStatusBar ? resize : "false"}
                    />
                )}
            </div>
            <ValidationAlert>{stringAttribute.validation}</ValidationAlert>
        </Fragment>
    );
}
