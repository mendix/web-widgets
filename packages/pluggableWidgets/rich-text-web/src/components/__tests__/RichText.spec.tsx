import { createElement } from "react";
import "@testing-library/jest-dom";
import { RichText } from "../RichText";
import { CKEditorConfig } from "ckeditor4-react";
import { getPreset, defineEnterMode, getToolbarGroupByName, defineAdvancedGroups } from "../../utils/ckeditorConfigs";
import { mount, ReactWrapper } from "enzyme";
import renderer from "react-test-renderer";
import { getDimensions, EditableValueBuilder } from "@mendix/pluggable-widgets-commons";
import { TOOLBAR_GROUP, ToolbarGroup } from "../../utils/ckeditorPresets";
import { AdvancedConfigType, RichTextContainerProps } from "../../../typings/RichTextProps";

const defaultRichTextProps: RichTextContainerProps = {
    stringAttribute: new EditableValueBuilder<string>().withValue("Rich text default value").build(),
    sanitizeContent: false,
    advancedContentFilter: "auto",
    name: "RichText_Div",
    spellChecker: false,
    readOnlyStyle: "text",
    editorType: "classic",
    enterMode: "paragraph",
    shiftEnterMode: "paragraph",
    tabIndex: 1,
    advancedMode: false,
    preset: "basic",
    widthUnit: "percentage",
    width: 100,
    heightUnit: "percentageOfWidth",
    height: 75,
    toolbarConfig: "basic",
    documentGroup: true,
    clipboardGroup: true,
    editingGroup: true,
    formsGroup: true,
    separatorGroup: true,
    basicStylesGroup: true,
    paragraphGroup: true,
    linksGroup: true,
    separator2Group: true,
    stylesGroup: true,
    colorsGroup: true,
    toolsGroup: true,
    othersGroup: true,
    advancedConfig: [],
    codeHighlight: false,
    allowedContent: "",
    disallowedContent: "",
    enableUploadImages: false,
    uploadImageEndpoint: "",
    id: "1.Dev.Test_ListenTo.richText1_x_1"
};

describe("RichText", () => {
    window.mx = {
        remoteUrl: "https://example.com",
        session: {
            getConfig: (value: string) => value + "config"
        }
    };

    function renderRichText(props = defaultRichTextProps): ReactWrapper {
        return mount(<RichText {...props} />);
    }

    it("render DOM structure", () => {
        const richText = renderer.create(<RichText {...defaultRichTextProps} />).toJSON();
        expect(richText).toMatchSnapshot();
    });

    it("renders dom elements with correct props", () => {
        const richText = renderRichText();
        expect(richText).toBeDefined();
        expect(richText.find(`#${defaultRichTextProps.name}`)).toBeDefined();

        const richTextElement = richText.find(".widget-rich-text");
        expect(richTextElement?.getElements().length).toBeGreaterThan(0);

        const styleProps = richTextElement.getElements()[0].props.style;
        const dimensions = getDimensions(defaultRichTextProps);
        expect(styleProps.width).toEqual(dimensions.width);
        expect(styleProps.height).toEqual(dimensions.height);

        expect(richText.find(`.editor-${defaultRichTextProps.readOnlyStyle}`)).toEqual({});
    });

    it("renders MainEditor component with correct props", () => {
        const richText = renderRichText();
        const mainEditor = richText.find("MainEditor");
        expect(mainEditor).toBeDefined();
        const props = mainEditor.props() as CKEditorConfig;
        const dimensions = getDimensions(defaultRichTextProps);
        const toolbar = getPreset("basic");
        expect(props.config.config).toMatchObject({
            autoGrow_minHeight: 300,
            toolbarCanCollapse: true,
            autoGrow_onStartup: true,
            width: dimensions.width,
            height: dimensions.height,
            enterMode: 1,
            shiftEnterMode: 1,
            disableNativeSpellChecker: !defaultRichTextProps.spellChecker,
            ...toolbar
        });
    });

    it("renders editor in read only mode with specified style", () => {
        const containerProps: RichTextContainerProps = {
            ...defaultRichTextProps,
            stringAttribute: new EditableValueBuilder<string>()
                .withValue("Rich text default value")
                .isReadOnly()
                .build()
        };
        const richText = renderRichText(containerProps);
        expect(richText.find(`.editor-${containerProps.readOnlyStyle}`).getElements().length).toBeGreaterThan(0);
    });
});

describe("CKEditor configuration", () => {
    it("returns correct value for enterMode", () => {
        let enterMode = defineEnterMode("paragraph");
        expect(enterMode).toEqual(1);

        enterMode = defineEnterMode("breakLines");
        expect(enterMode).toEqual(2);

        enterMode = defineEnterMode("blocks");
        expect(enterMode).toEqual(3);

        enterMode = defineEnterMode("");
        expect(enterMode).toEqual(1);
    });

    it("returns group named document", () => {
        const group = getToolbarGroupByName("document");
        expect(group).toEqual(TOOLBAR_GROUP[0]);
    });

    it("creates toolbar from advanced configuration", () => {
        const input: AdvancedConfigType[] = [
            { ctItemType: "About", ctItemToolbar: "toolbar1" },
            { ctItemType: "Anchor", ctItemToolbar: "toolbar2" }
        ];
        const grouped = defineAdvancedGroups({ ...defaultRichTextProps, advancedConfig: input });
        const result = [
            {
                name: "toolbar1",
                items: ["About"]
            },
            {
                name: "toolbar2",
                items: ["Anchor"]
            }
        ];
        expect(grouped).toMatchObject(result);
    });

    it("returns configuration object for basic preset", () => {
        const preset = "basic";
        const config = getPreset(preset);

        expect(config).toEqual({
            toolbarGroups: TOOLBAR_GROUP,
            removeButtons:
                "Source,Save,Templates,NewPage,ExportPdf,Preview,Print,Cut,Redo,Copy,Undo," +
                "Paste,PasteText,PasteFromWord,Find,Replace,SelectAll,Scayt," +
                "Form,Checkbox,Radio,TextField,Textarea,Select," +
                "Button,ImageButton,HiddenField,Strike,Underline,Subscript," +
                "Superscript,CopyFormatting,RemoveFormat,CreateDiv,Blockquote,JustifyLeft," +
                "BidiLtr,Image,Table,BidiRtl,JustifyCenter,JustifyRight,Language,JustifyBlock," +
                "HorizontalRule,SpecialChar,Smiley,PageBreak,Iframe,Styles,TextColor,BGColor," +
                "ShowBlocks,Maximize,Format,Font,FontSize,Anchor"
        });
    });

    it("returns configuration object for standard preset", () => {
        const preset = "standard";
        const config = getPreset(preset);

        const toolbarGroups: Array<ToolbarGroup | string> = [...TOOLBAR_GROUP];
        toolbarGroups.splice(4, 0, "/");
        expect(config).toEqual({
            toolbarGroups,
            extraPlugins: "mxupload",
            removeButtons:
                "Save,Templates,NewPage,ExportPdf,Preview,Print,Find," +
                "Replace,SelectAll,Form,Checkbox,Radio,TextField,Textarea," +
                "Select,Button,ImageButton,HiddenField,Underline,Subscript," +
                "Superscript,CopyFormatting,CreateDiv,JustifyLeft,BidiLtr,BidiRtl," +
                "JustifyCenter,JustifyRight,Language,JustifyBlock,Smiley,PageBreak," +
                "Iframe,TextColor,BGColor,Font,FontSize,ShowBlocks"
        });
    });
});
