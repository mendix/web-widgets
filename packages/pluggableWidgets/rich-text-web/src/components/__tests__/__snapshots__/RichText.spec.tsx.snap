// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RichText render DOM structure 1`] = `
<div
  className="widget-rich-text"
  id="1.Dev.Test_ListenTo.richText1_x_1"
  style={
    Object {
      "height": "auto",
      "width": "100%",
    }
  }
>
  <div
    id="RichText_Div"
  />
</div>
`;
